File: Java/src/MergeSort.java

public class MergeSort {

	public static void main(String[] args) {
		// especial para arrays
		int arr[]= {13,139,2642,186,785,-1000000,-634,7,318};
		print(arr);
		sort(arr, 0, arr.length-1);
		print(arr);
	}
	
	static void print(int arr[]) {
		for (int i = 0; i < arr.length; i++) {
			System.out.print(arr[i] +" ");
		}
		System.out.println();
	}
	
	static void merge (int arr[], int left, int mid, int right) {
		int n1=mid-left+1;
		int n2=right-mid;
		int L[]= new int[n1];
		print(L);
		int R[]=new int[n2];
		print(R);
		for (int i = 0; i < n1; i++) {
			L[i]=arr[left+i];
		}
		for (int i = 0; i < n2; i++) {
			R[i]=arr[mid+1+i];
		}
		int i=0,j=0;
		int k=left;
		while (i<n1&& j<n2) {
			if (L[i]<=R[j]) {
				arr[k]=L[i];
				i++;
			}else {
				arr[k]=R[j];
				j++;
			}
			k++;
		}
		while (i<n1) {
			arr[k]=L[i];
			i++;
			k++;
		}
		while (j<n2) {
			arr[k]=R[j];
			j++;
			k++;
		}
		
		
	}
	
	static void sort(int arr[],int left,int right) {
		if (left<right) {
			int mid=(left+right)/2;
			sort(arr, left, mid);
			sort(arr, mid+1, right);
			merge(arr, left,mid,right);
		}
	}

}

File: Codigos_por_competidor/Diego_Rodriguez/Algoritmos/Merge_Sort.kt
import kotlin.jvm.JvmStatic
import java.io.BufferedReader
import java.io.InputStreamReader
import kotlin.Throws
import java.io.IOException

object Merge_Sort {
    @Throws(IOException::class)
    @JvmStatic
    fun main(args: Array<String>) {
        val br = BufferedReader(InputStreamReader(System.`in`))
        val entrada = br.readLine().split(" ").toTypedArray()
        val arr = IntArray(entrada.size)
        for (i in entrada.indices) {
            arr[i] = entrada[i].toInt()
        }
        mergeSort(arr, 0, arr.size - 1)
        print(arr)
    }

    fun merge(arr: IntArray, left: Int, mid: Int, right: Int) {
        val tam1 = mid - left + 1
        val tam2 = right - mid
        val Left = IntArray(tam1)
        val Right = IntArray(tam2)
        for (i in 0 until tam1) {
            Left[i] = arr[left + i]
        }
        for (i in 0 until tam2) {
            Right[i] = arr[mid + 1 + i]
        }
        var i = 0
        var j = 0
        var k = left
        while (i < tam1 && j < tam2) {
            if (Left[i] <= Right[j]) {
                arr[k] = Left[i]
                i++
            } else {
                arr[k] = Right[j]
                j++
            }
            k++
        }
        while (i < tam1) {
            arr[k] = Left[i]
            i++
            k++
        }
        while (j < tam2) {
            arr[k] = Right[j]
            j++
            k++
        }
    }

    fun mergeSort(arr: IntArray, left: Int, right: Int) {
        if (left < right) {
            val m = (left + right) / 2
            mergeSort(arr, left, m)
            mergeSort(arr, m + 1, right)
            merge(arr, left, m, right)
        }
    }

    fun print(arr: IntArray) {
        val n = arr.size
        for (i in 0 until n) {
            print(arr[i].toString() + " ")
        }
        println()
    }
}